p2p/src/behaviour/firewall/permissions.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+pub use stronghold_derive::RequestPermissions;
+
+/// The permission value for request variants.
+/// This is realized as a bit set at a certain index, hence the value is always a power of 2.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct PermissionValue(u32);
+
+impl PermissionValue {
```

Comment:
This could be generic, that `PermissionValue` would have generic Type `T`, that is being comparable and hashable. This way we can keep the number of permissions dynamic. I can't see a reason why the permissions have a purely numeric value that is being stored inside an unsigned int. Does the inherent ordering have a semantic value inside the firewall rules?

Reply:
In general, the logic inside of this `permissions.rs` module is not used anymore in the stronghold-p2p library itself. Instead it provides some logic for convenience, that may be used by the user when writing a `Rule::Restricted. An example for this implemented in the test `test_firewall::firewall_permissions`.

But it being a numeric value is on purpose yes. The idea is (similar to how the `rwx` permissions in unix systems implement it) that each variant has a numeric value to the power of 2, e.g. ReadStore: 1, WriteStore: 2, WriteVault: 4, ExecuteProcedure: 8, ..., so that the a `FirewallPermission` can be created from the sum of allowed `PermissionValue`s.
It could be implemented that the type of integer is generic, but I decided against it because 1. It's unlikely that there are request messages enum with more than 32 variants , and 2. the user can always do their own implementation if this doesn't fit their use-case.